### PR TITLE
Adds application APIs

### PR DESCRIPTION
**Pull request by robertjd**
* The "auto generated" header in all files has been modified to use a different syntax that does not confuse the JsDoc parser
* Factory resolution maps now reference constructors or factories, rather than strings which indicate a model class.  This allows for nested class discrimination
* JsDoc has been fixed to properly indicate that Client extends GeneratedClient (not the other way around)
* Model properties are now enumerated as JsDoc tags on the class
* Model properties of a known class are now new'd up as that class
* The Application class methods are tested in addition to the underlying client methods
* The README has been updated with Application information and some other cleanup
* The JsDoc home page (a markdown file) is now auto-generated by parsing the root README and modifying the few pieces that we care about.

### CODE REVIEW

File: test/it/client-create-application.js [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+  });
+
+  it('should allow me to create a OIDC client application', async () => {
+    const application = {
+      name: 'oidc_client',
+      label: 'Sample Client',
+      signOnMode: 'OPENID_CONNECT',
+      credentials: {
+        oauthClient: {
+          client_id: '0oa1hm4POxgJM6CPu0g4',
```
Comment (lboyette-okta) [inline test/it/client-create-application.js]: we allow defining your own client_id?

File: test/it/client-delete-application-user.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      await utils.removeAppByLabel(client, application.label);
+      await utils.cleanup(client, user);
+      createdApplication = await client.createApplication(application);
+      createdUser = await client.createUser(user);
+      createdAppUser = await client.assignUserToApplication(createdApplication.id, {
+        id: createdUser.id
+      });
+      await client.deleteApplicationUser(createdApplication.id, createdAppUser.id)
+      .then(result => {
+        expect(result.status).to.equal(204);
```
Comment (lboyette-okta) [inline test/it/client-delete-application-user.js]: I'm trying to remember, do we only return this status property on delete? I thought there was some nesting ie. `result.httpResponse.status`

File: README.md [hunk 3]
```diff
@@ -31,20 +31,33 @@ const client = new okta.Client({
 
 It is also possible to provide configuration through environment variables or YAML files.  Please see [Configuration](#configuration) for examples.
 
+All interaction with the [Okta Platform API] is done through client methods.  Some examples are below, but for a full list of methods please refer to the JsDoc page for the client:
+
+https://developer.okta.com/okta-sdk-nodejs/jsdocs/Client.html
+
 ## Examples
 
 This libray is a wrapper for the [Okta Platform API], which should be referred to as the source-of-truth for what is and isn't possible with the API.  In the following sections we show you how to use your client to perform some common operations with the [Okta Platform API].
```
Comment (vijetmahabaleshwar-okta) [inline README.md]: libray -> library

File: src/generated-client.js [hunk 4]
```diff
@@ -27,10 +24,344 @@ const ModelFactory = require('./model-factory');
  * Auto-Generated API client, implementes the operations as defined in the OpenaAPI JSON spec
```
Comment (vijetmahabaleshwar-okta) [inline src/generated-client.js]: implementes -> implements

File: templates/generated-client.js.hbs [hunk 5]
```diff
@@ -14,7 +11,6 @@ const ModelFactory = require('./model-factory');
  * Auto-Generated API client, implementes the operations as defined in the OpenaAPI JSON spec
  *
  * @class GeneratedApiClient
- * @extends {Client}
```
[the remark this answers is not in the document]
Comment (robertjd) [reply templates/generated-client.js.hbs]: It doesn't, it's the other way around (so this annotation was wrong)

File: test/it/client-delete-application-user.js (hunk 2 above)
Comment (robertjd) [reply test/it/client-delete-application-user.js]: Yeah, right now we just return the http response for delete operations.  It's not consistent with other operations, and I want to change this (at which point these tests will also have to change)

File: test/it/client-create-application.js (hunk 1 above)
Comment (robertjd) [reply test/it/client-create-application.js]: yep

File: templates/index.js [hunk 6]
```diff
@@ -104,6 +105,34 @@ js.process = ({spec, operations, models, handlebars}) => {
...
+  });
+
+  handlebars.registerHelper('propertyCastBuilder', (model) => {
+    if (!model.properties) {
+      return;
+    }
+    const constructorStatements = [];
+    model.properties.forEach(property => {
+      if (property.$ref && property.model !== 'object' && !property.isEnum) {
+        constructorStatements.push(`    if (resourceJson && resourceJson.${property.propertyName}) {`);
```
[the remark this answers is not in the document]
Comment (robertjd) [reply templates/index.js]: Yeah there's a lot of ugly hacks that i'd like to cleanup at some point.  I looked at common-tags, not sure if i like it much better?

File: utils/make-jsdoc-readme.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#! /usr/bin/env node
+
+/**
+ * This file transforms the README into another markdown file that is used as the home page for the JsDoc site
+ *
+ * The things we want to transform are:
+ *
+ * - Removing the table-of-contents for the examples, because the JsDoc site has it's own TOC sidebar
```
Comment (lboyette-okta) [inline utils/make-jsdoc-readme.js]: nit: its

**Pull request merged**